File: src/main/scala/cosim/EmuTestResultStage.scala
// Copyright (c) 2018 Norwegian University of Science and Technology (NTNU)
// Copyright (c) 2019 Xilinx
//
// BSD v3 License
//
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
// * Redistributions of source code must retain the above copyright notice, this
//   list of conditions and the following disclaimer.
//
// * Redistributions in binary form must reproduce the above copyright notice,
//   this list of conditions and the following disclaimer in the documentation
//   and/or other materials provided with the distribution.
//
// * Neither the name of [project] nor the names of its
//   contributors may be used to endorse or promote products derived from
//   this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

package bismo

import Chisel._
import fpgatidbits.dma._
import fpgatidbits.ocm._
import fpgatidbits.streams._
import fpgatidbits.PlatformWrapper._

class EmuTestResultStage(
  accArrayDim: Int, p: PlatformWrapperParams) extends GenericAccelerator(p) {
  val numMemPorts = 1
  // parameters for accelerator instance
  val myP = new ResultStageParams(
    accWidth = 32, resMemReadLatency = 0,
    dpa_rhs = accArrayDim, dpa_lhs = accArrayDim, mrp = PYNQZ1Params.toMemReqParams())
  val io = new GenericAcceleratorIF(numMemPorts, p) {
    // base control signals
    val start = Bool(INPUT)                   // hold high while running
    val done = Bool(OUTPUT)                   // high when done until start=0
    val csr = new ResultStageCtrlIO().asInput
    val accwr_en = Bool(INPUT)
    val accwr_lhs = UInt(INPUT, width = log2Up(myP.dpa_lhs))
    val accwr_rhs = UInt(INPUT, width = log2Up(myP.dpa_rhs))
    val accwr_data = UInt(INPUT, width = myP.accWidth)
  }
  val resmem = Vec.fill(myP.dpa_lhs) {
    Vec.fill(myP.dpa_rhs) {
      Module(new PipelinedDualPortBRAM(
        addrBits = 1, dataBits = myP.accWidth, regIn = 0, regOut = 0)).io
    }
  }
  val res = Module(new ResultStage(myP)).io
  res.start := io.start
  io.done := res.done
  res.csr <> io.csr
  for (lhs ← 0 until myP.dpa_lhs) {
    for (rhs ← 0 until myP.dpa_rhs) {
      // drive defaults on resmem req port 0
      val is_my_lhs = (UInt(lhs) === io.accwr_lhs)
      val is_my_rhs = (UInt(rhs) === io.accwr_rhs)
      val is_my_write = is_my_lhs & is_my_rhs
      // write enable selected resmem entry
      resmem(lhs)(rhs).ports(0).req.writeEn := is_my_write & io.accwr_en
      resmem(lhs)(rhs).ports(0).req.addr := UInt(0)
      resmem(lhs)(rhs).ports(0).req.writeData := io.accwr_data
      // connect resmem port 1 directly to ResultStage interface
      res.resmem_req(lhs)(rhs) <> resmem(lhs)(rhs).ports(1).req
      resmem(lhs)(rhs).ports(1).rsp <> res.resmem_rsp(lhs)(rhs)
    }
  }

  // connect DRAM interface for ResultStage
  res.dram.wr_req <> io.memPort(0).memWrReq
  res.dram.wr_dat <> io.memPort(0).memWrDat
  io.memPort(0).memWrRsp <> res.dram.wr_rsp
  // plug unused read port
  plugMemReadPort(0)
  // the signature can be e.g. used for checking that the accelerator has the
  // correct version. here the signature is regenerated from the current date.
  io.signature := makeDefaultSignature()
}

File: src/test/cosim/EmuTestResultStage.cpp
// Copyright (c) 2018 Norwegian University of Science and Technology (NTNU)
// Copyright (c) 2019 Xilinx
//
// BSD v3 License
//
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
// * Redistributions of source code must retain the above copyright notice, this
//   list of conditions and the following disclaimer.
//
// * Redistributions in binary form must reproduce the above copyright notice,
//   this list of conditions and the following disclaimer in the documentation
//   and/or other materials provided with the distribution.
//
// * Neither the name of BISMO nor the names of its
//   contributors may be used to endorse or promote products derived from
//   this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

#include <cassert>
#include <iostream>
using namespace std;
#include "platform.h"
#include "EmuTestResultStage.hpp"

// Cosim test for basic ResultStage behavior

#define DPA_LHS   2
#define DPA_RHS   2
typedef int32_t   ResultType;

WrapperRegDriver * p;
EmuTestResultStage * dut;

void exec_and_wait() {
  dut->set_start(1);
  while(dut->get_done() != 1);
  dut->set_start(0);
}

// get offset of element (x, y) in a 1D array[nx][ny]
int offset2D(int x, int y, int ny) {
  return x * ny + y;
}

void do_result_write(void * accel_buf, int ind_rhs, int ind_lhs, int num_rhs, int num_lhs) {
  size_t stride = num_lhs * sizeof(ResultType);
  uint64_t base = sizeof(int32_t) * offset2D(ind_rhs, ind_lhs, num_lhs) + (uint64_t) accel_buf;
  dut->set_csr_dram_base(base);
  dut->set_csr_dram_skip(stride);
  dut->set_csr_waitComplete(0);
  dut->set_csr_waitCompleteBytes(0);
  dut->set_csr_resmem_addr(0);
  exec_and_wait();
}

void do_result_waitcomplete(size_t nbytes) {
  dut->set_csr_waitComplete(1);
  dut->set_csr_waitCompleteBytes(nbytes);
  exec_and_wait();
}

void write_resmem(int lhs, int rhs, int val) {
  dut->set_accwr_lhs(lhs);
  dut->set_accwr_rhs(rhs);
  dut->set_accwr_data(val);
  dut->set_accwr_en(1);
  dut->set_accwr_en(0);
}

int main()
{
  bool all_OK = true;
  p = initPlatform();
  dut = new EmuTestResultStage(p);

  bool simple_res_OK = true;
  const size_t nrhs = DPA_RHS * 1, nlhs = DPA_LHS * 2;
  size_t nbytes = sizeof(ResultType) * nrhs * nlhs;
  uint32_t * hostbuf = new uint32_t[nrhs * nlhs];
  void * accelbuf = p->allocAccelBuffer(nbytes);

  // each tile is written to memory in column-major order, with appropriate
  // stride to account for the tiling. e.g the 2x2 DPA array:
  //   (<---rhs dim--->)       ^
  //          A B            lhs dim
  //          C D              v
  // turns into:
  // A, C, [stride], B, D
  // when written into memory, so the lhs entires remain contiguous, "rhs-major"

  //dut->set_acc_in_0_0(1);
  write_resmem(0, 0, 1);
  //dut->set_acc_in_1_0(2);
  write_resmem(1, 0, 2);
  //dut->set_acc_in_0_1(5);
  write_resmem(0, 1, 5);
  //dut->set_acc_in_1_1(6);
  write_resmem(1, 1, 6);
  do_result_write(accelbuf, 0 * DPA_RHS, 0 * DPA_LHS, nrhs, nlhs);

  //dut->set_acc_in_0_0(3);
  write_resmem(0, 0, 3);
  //dut->set_acc_in_1_0(4);
  write_resmem(1, 0, 4);
  //dut->set_acc_in_0_1(7);
  write_resmem(0, 1, 7);
  //dut->set_acc_in_1_1(8);
  write_resmem(1, 1, 8);
  do_result_write(accelbuf, 0 * DPA_RHS, 1 * DPA_LHS, nrhs, nlhs);

  // wait until all writes are completed
  do_result_waitcomplete(nbytes);

  p->copyBufferAccelToHost(accelbuf, hostbuf, nbytes);

  for(int i = 0; i < nrhs * nlhs; i++) {
    simple_res_OK &= (hostbuf[i] == i+1);
  }
  cout << "Simple result write test passed? " << simple_res_OK << endl;
  if(!simple_res_OK) {
    for(int i = 0; i < nrhs * nlhs; i++) {
      cout << "i[" << i << "] = " << hostbuf[i] << endl;
    }
  }
  all_OK &= simple_res_OK;

  delete [] hostbuf;

  delete dut;
  deinitPlatform(p);

  return all_OK ? 0 : -1;
}
